<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU18_Infraestructuras_eventos</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 04. Asignar distrito censal a infraestructuras

* Dadas las coordenadas de cada infraestructura, se asigna el distrito censal al que pertenece.







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos


In [39]:
library(sf)
library(readr)
library(dplyr)


### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

#### 1. Archivo de infraestructuras

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [3]:
iFile_01 <- "CU_18_05_02_infraestructuras.csv"
file_data_01 <- paste0(iPath, iFile_01)

if(file.exists(file_data_01)){
    cat("Se leerán datos del archivo: ", file_data_01)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_18_05_02_infraestructuras.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [4]:
data_01 <- read_csv(file_data_01)


Rows: 1633 Columns: 8
-- Column specification --------------------------------------------------------
Delimiter: ","
chr (6): grupo, tipo, nombre, CODMUN, DIRECCION, info
dbl (2): X, Y

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [5]:
glimpse(data_01)

Rows: 1,633
Columns: 8
$ grupo     <chr> "Transporte", "Transporte", "Transporte", "Transporte", "Tra~
$ tipo      <chr> "Intercambiadores", "Intercambiadores", "Intercambiadores", ~
$ nombre    <chr> "Grandes Intercambiadores Plaza El<U+00ED>ptica", "Grandes Intercam~
$ CODMUN    <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079~
$ DIRECCION <chr> "Plaza El<U+00ED>ptica s/n", "Calle Princesa, 89", "Estaci<U+00F3>n Pr<U+00ED>nci~
$ X         <dbl> -3.716577, -3.719508, -3.719560, -3.689044, -3.676731, -3.68~
$ Y         <dbl> 40.38540, 40.43474, 40.42080, 40.46719, 40.43797, 40.47198, ~
$ info      <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~


Muestra de datos:

In [6]:
slice_head(data_01, n = 5)

grupo      tipo            
1 Transporte Intercambiadores
2 Transporte Intercambiadores
3 Transporte Intercambiadores
4 Transporte Intercambiadores
5 Transporte Intercambiadores
  nombre                                              CODMUN
1 Grandes Intercambiadores Plaza El<U+00ED>ptica      079   
2 Grandes Intercambiadores Moncloa                    079   
3 Grandes Intercambiadores Pr<U+00ED>ncipe P<U+00ED>o 079   
4 Grandes Intercambiadores Plaza de Castilla          079   
5 Grandes Intercambiadores Avenida de Am<U+00E9>rica  079   
  DIRECCION                                  X         Y        info
1 Plaza El<U+00ED>ptica s/n                  -3.716577 40.38540 NA  
2 Calle Princesa, 89                         -3.719508 40.43474 NA  
3 Estaci<U+00F3>n Pr<U+00ED>ncipe P<U+00ED>o -3.719560 40.42080 NA  
4 Calle San Aquilino y Avenida de Asturias   -3.689044 40.46719 NA  
5 Avenida de Am<U+00E9>rica 9 A              -3.676731 40.43797 NA

#### 2. Archivo de geolocalización distritos censales

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data_02 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [18]:
iFile_02 <- "CU_18_05_03_distritos_geo.json"
file_data_02 <- paste0(iPath, iFile_02)

if(file.exists(file_data_02)){
    cat("Se leerán datos del archivo: ", file_data_02)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_18_05_03_distritos_geo.json

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [23]:
data_02 <- st_read(file_data_02)


Reading layer `CU_18_05_03_distritos_geo' from data source 
  `/Users/emilio.lcano/academico/gh_repos/__transferencia/citizenlab/CitizenLab-Research-and-Development/casos_urjc/notebooks/II_data_processing/18_infraestructuras/Data/Input/CU_18_05_03_distritos_geo.json' 
  using driver `GeoJSON'
Simple feature collection with 247 features and 2 fields
Geometry type: GEOMETRY
Dimension:     XY
Bounding box:  xmin: -4.579006 ymin: 39.8848 xmax: -3.052983 ymax: 41.16584
Geodetic CRS:  WGS 84


Estructura de  los datos:

In [10]:
glimpse(data_02)

Rows: 247
Columns: 3
$ CMUN     <chr> "001", "002", "003", "004", "005", "005", "005", "005", "005"~
$ CDIS     <chr> "01", "01", "01", "01", "01", "02", "03", "04", "05", "01", "~
$ geometry <POLYGON [arc_degree]> POLYGON ((-3.64502 41.12129..., POLYGON ((-3~


Muestra de datos:

In [11]:
slice_head(data_02, n = 5)

Registered S3 method overwritten by 'geojsonsf':
  method        from   
  print.geojson geojson



CMUN CDIS geometry                      
1 001  01   POLYGON ((-3.64502 41.12129...
2 002  01   POLYGON ((-3.503032 40.526,...
3 003  01   POLYGON ((-3.808664 40.8921...
4 004  01   POLYGON ((-4.00197 40.25642...
5 005  01   POLYGON ((-3.361691 40.4762...

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Archivo de POIs infraestructuras
* Archivo de georeferenciación distritos censales

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

#### Data Transform


Convertir coordenadas numéricas a objeto espacial

In [12]:
tdata_01 <- data_01 |>
    st_as_sf(coords = c("X", "Y"), crs = 4326,
           remove = FALSE) 

In [13]:
slice_head(tdata_01, n = 5)

grupo      tipo            
1 Transporte Intercambiadores
2 Transporte Intercambiadores
3 Transporte Intercambiadores
4 Transporte Intercambiadores
5 Transporte Intercambiadores
  nombre                                              CODMUN
1 Grandes Intercambiadores Plaza El<U+00ED>ptica      079   
2 Grandes Intercambiadores Moncloa                    079   
3 Grandes Intercambiadores Pr<U+00ED>ncipe P<U+00ED>o 079   
4 Grandes Intercambiadores Plaza de Castilla          079   
5 Grandes Intercambiadores Avenida de Am<U+00E9>rica  079   
  DIRECCION                                  X         Y        info
1 Plaza El<U+00ED>ptica s/n                  -3.716577 40.38540 NA  
2 Calle Princesa, 89                         -3.719508 40.43474 NA  
3 Estaci<U+00F3>n Pr<U+00ED>ncipe P<U+00ED>o -3.719560 40.42080 NA  
4 Calle San Aquilino y Avenida de Asturias   -3.689044 40.46719 NA  
5 Avenida de Am<U+00E9>rica 9 A              -3.676731 40.43797 NA  
  geometry                  
1 POINT (-3.716577 40.3854) 
2 POINT (-3.719508 40.43474)
3 POINT (-3.71956 40.4208)  
4 POINT (-3.689044 40.46719)
5 POINT (-3.676731 40.43797)

* Corregir geometrías erróneas de distritos


In [27]:
tdata_02 <- data_02 |> st_make_valid()


* Unir data frames de POIs y distritos por intersección de coordenadas

In [28]:
data <- tdata_01 |> 
  st_join(tdata_02) |> 
  st_drop_geometry()

In [30]:
data |> slice_head(n = 5)

grupo      tipo            
1 Transporte Intercambiadores
2 Transporte Intercambiadores
3 Transporte Intercambiadores
4 Transporte Intercambiadores
5 Transporte Intercambiadores
  nombre                                              CODMUN
1 Grandes Intercambiadores Plaza El<U+00ED>ptica      079   
2 Grandes Intercambiadores Moncloa                    079   
3 Grandes Intercambiadores Pr<U+00ED>ncipe P<U+00ED>o 079   
4 Grandes Intercambiadores Plaza de Castilla          079   
5 Grandes Intercambiadores Avenida de Am<U+00E9>rica  079   
  DIRECCION                                  X         Y        info CMUN CDIS
1 Plaza El<U+00ED>ptica s/n                  -3.716577 40.38540 NA   079  12  
2 Calle Princesa, 89                         -3.719508 40.43474 NA   079  09  
3 Estaci<U+00F3>n Pr<U+00ED>ncipe P<U+00ED>o -3.719560 40.42080 NA   079  09  
4 Calle San Aquilino y Avenida de Asturias   -3.689044 40.46719 NA   079  05  
5 Avenida de Am<U+00E9>rica 9 A              -3.676731 40.43797 NA   079  05

Si no aplica: Estos datos no requieren tareas de este tipo.

## Synthetic Data Generation

Si no aplica: Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Si no aplica: Estos datos no requieren tareas de este tipo.

## Open Data

Los datos provienen de fuentes abiertas obtenidas en taras anteriores.

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [ ]:
data_to_save <- data


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [31]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_04"
archivo <- ""
proper <- "_infraestructuras"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(data, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [33]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
write_csv(data, path_out)

cat('File saved as: ')
path_out

File saved as: 

[1] "Data/Output/CU_18_05_04_infraestructuras.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [34]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {readr}, {dplyr} and {sf} packages must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* dplyr 1.0.10
* sf 1.0.9


### Data structures

#### Objeto `data`

- Los datos de origen contienen datos de infraesctructuras
- Hay 1633 filas con las variables: 
    * grupo
	* tipo
	* nombre
	* CODMUN
	* DIRECCION
	* X
	* Y
	* info
	* CMUN
	* CDIS


In [35]:
glimpse(data)

Rows: 1,633
Columns: 10
$ grupo     <chr> "Transporte", "Transporte", "Transporte", "Transporte", "Tra~
$ tipo      <chr> "Intercambiadores", "Intercambiadores", "Intercambiadores", ~
$ nombre    <chr> "Grandes Intercambiadores Plaza El<U+00ED>ptica", "Grandes Intercam~
$ CODMUN    <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079~
$ DIRECCION <chr> "Plaza El<U+00ED>ptica s/n", "Calle Princesa, 89", "Estaci<U+00F3>n Pr<U+00ED>nci~
$ X         <dbl> -3.716577, -3.719508, -3.719560, -3.689044, -3.676731, -3.68~
$ Y         <dbl> 40.38540, 40.43474, 40.42080, 40.46719, 40.43797, 40.47198, ~
$ info      <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
$ CMUN      <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079~
$ CDIS      <chr> "12", "09", "09", "05", "05", "05", "06", "02", "01", "10", ~


#### Observaciones generales sobre los datos

* Hay algunas discrepancias entre los códigos de municipios de las infraestructuras y los de los distritos censales que habría que revisar
* En lo sucesivo se tomarán como buenos los códigos del INE al ser la fuente primaria

### Consideraciones para despliegue en piloto

* xxx

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han transformado las coordenadas numéricas de infraestructuras a objeto espacial `sf``
- Se han corregido las geometrías erróneas de distritos
- Se ha obtenido el distrito al que pertenece la infraestructura por intersección espacial


#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se debe revisar la codificación de municipios de la Cumunidad de Madrid

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código